### PR TITLE
code update by proposal

### PR DESCRIPTION
**Pull request by scf0220**
- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/okex/okexchain/blob/develop/CONTRIBUTING.md#pr-targeting))

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge PR #XYZ: [title]" ([coding standards](https://github.com/okex/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/gov/ante/ante.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	sk stakingkeeper.Keeper
+	ak auth.AccountKeeper
+}
+
+func NewAnteDecorator(k stakingkeeper.Keeper, ak auth.AccountKeeper) AnteDecorator {
+	return AnteDecorator{sk: k, ak: ak}
+}
+
+func (ad AnteDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	m := tx.GetMsgs()[0]
```
Comment (BananaLF) [inline x/gov/ante/ante.go]: you need check msgs

File: x/gov/ante/ante.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	switch msg := m.(type) {
+	case types.MsgSubmitProposal:
+		switch proposalType := msg.Content.(type) {
+		case evmtypes.ManagerContractByteCodeProposal:
+			if !ad.sk.IsValidator(ctx, msg.GetSigners()[0]) {
+				return ctx, evmtypes.ErrCodeProposerMustBeValidator()
+			}
+
+			// check operation contract
+			contract := ad.ak.GetAccount(ctx, proposalType.Contract)
```
Comment (BananaLF) [inline x/gov/ante/ante.go]: i think check contract and substitute_contract is not need.

File: x/gov/ante/ante.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+	return AnteDecorator{sk: k, ak: ak}
+}
+
+func (ad AnteDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	m := tx.GetMsgs()[0]
+	switch msg := m.(type) {
+	case types.MsgSubmitProposal:
+		switch proposalType := msg.Content.(type) {
+		case evmtypes.ManagerContractByteCodeProposal:
+			if !ad.sk.IsValidator(ctx, msg.GetSigners()[0]) {
```
Comment (BananaLF) [inline x/gov/ante/ante.go]: use msg.proposer much better

File: x/gov/ante/ante.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+				return ctx, evmtypes.ErrCodeProposerMustBeValidator()
+			}
+
+			// check operation contract
+			contract := ad.ak.GetAccount(ctx, proposalType.Contract)
+			contractAcc, ok := contract.(*ethermint.EthAccount)
```
Comment (BananaLF) [inline x/gov/ante/ante.go]: check contract == nil much better

File: x/evm/keeper/code_update.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+// UpdateContractBytecode update contract bytecode
+func (k *Keeper) UpdateContractBytecode(ctx sdk.Context, p types.ManagerContractByteCodeProposal) sdk.Error {
+	contract := ethcmn.BytesToAddress(p.Contract)
+	substituteContract := ethcmn.BytesToAddress(p.SubstituteContract)
+
+	revertContractByteCode := p.Contract.String() == p.SubstituteContract.String()
+
+	preCode := k.EvmStateDb.GetCode(contract)
+	ContractAddr := k.EvmStateDb.GetAccount(contract)
```
Comment (BananaLF) [inline x/evm/keeper/code_update.go]: contractAcc

File: x/evm/keeper/code_update.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+		ctx.GetWatcher().SaveAccount(contractAfterUpdateCode)
+		return true
+	})
+	return nil
+}
+
+func (k *Keeper) storeInitContractCode(ctx sdk.Context, addr sdk.AccAddress, code []byte) {
+	store := k.paramSpace.CustomKVStore(ctx)
+	key := types.GetInitContractCode(addr)
+	if len(store.Get(key)) == 0 {
```
Comment (BananaLF) [inline x/evm/keeper/code_update.go]: use has much better

File: x/evm/types/key.go [hunk 7]
```diff
@@ -79,3 +80,7 @@ func splitBlockedContractAddress(key []byte) sdk.AccAddress {
 func GetSysContractAddressKey() []byte {
 	return append(KeyPrefixSysContractAddress, []byte(SysContractAddressKey)...)
 }
+
+func GetInitContractCode(contractAddr sdk.AccAddress) []byte {
```
Comment (BananaLF) [inline x/evm/types/key.go]: GetInitContractCodeKey much better

File: x/evm/keeper/code_update.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+	if ContractAddr == nil {
+		return types.ErrNotContracAddress(fmt.Errorf("%s", contract.String()))
+	}
+	preCodeHash := ContractAddr.CodeHash
+
+	var newCode []byte
+	if revertContractByteCode {
+		newCode = k.getInitContractCode(ctx, p.Contract)
+		if len(newCode) == 0 {
+			return nil
```
Comment (BananaLF) [inline x/evm/keeper/code_update.go]: may be return err much better

File: x/evm/keeper/code_update.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+		}
+	} else {
+		newCode = k.EvmStateDb.GetCode(substituteContract)
+	}
+	// update code
+	k.EvmStateDb.SetCode(contract, newCode)
```
Comment (BananaLF) [inline x/evm/keeper/code_update.go]: suggest not to use EvmStateDb, this create by mpt_pr. mpt_pr disable forever. I'm not sure EvmStateDb weather exist others data, so use CreateEmptyCommitStateDB

**Review by BananaLF — DISMISSED** [review]

**Review by ilovers — APPROVED** [review]

**Pull request merged**